### ISSUE TITLE
automatic application with fix_includes.py breaks order of preprocessor canstant definition and usage

### ISSUE THREAD

**Issue by pseyfert** [include-what-you-use/include-what-you-use#246]
in [this commit](https://github.com/pseyfert/root/commit/ebb32e3dae44f1e0cc6892268e2bbc2527ffcbb7) i fixed the few compiler errors introduced by running fix_includes.py. As you see, iwyu correctly identified that ROOT_VERSION is defined in RVersion.h and includes it, but it places the include after the usage of this preprocessor constant.

**Comment by vsapsai** [comment]
I see putting `#include "RVersion.h"` after `ROOT_VERSION_CODE` was done in [this commit](https://github.com/pseyfert/root/commit/2c0be929163d5140e66debb02f81ef412a60df03). My guess is that it's not really `fix_includes.py` bug, IWYU just doesn't detect dependencies between files, as far as I remember.

**Comment by pseyfert** [comment]
indeed, checking the dependencies between files is a feature i didn't really expect. though here it's not a dependency between different includes, but between an include and a line in the code. (though the longer i think about it, the harder it seems to me figuring out automatically where to place the includes)

**Comment by karies** [comment]
(While we can totally fix this on our side...): IWYU is supposed to detect `#include`s versus usages, and the use of a preproc macro is a use. I'm sure IWYU is detecting this (and thus including the header). Then it's a question of detecting the order: as IWYU knows what's triggering the need, it should be possible to include the headers before.

The problematic cases are CPP macros that influence subsequent includes - that's a mess and I'd say a fair limitation of IWYU :-)
